### PR TITLE
fix bugs in _get_xrefs and disassemble and so on

### PR DESCRIPTION
**Pull request by liumuqing**
1. It fix a bug when `src` or `dst` is `0` in _get_xrefs
2. It enable python-idb to disassemble arm, mips instruction.
3. It change the name of attribute `endEA` to `lastInstEA` of BasicBlock, to avoid confusion
4. Change some API to return None but not to raise an exception.

### CODE REVIEW

File: idb/analysis.py [hunk 1]
```diff
@@ -894,7 +894,10 @@ def get_chunks(self):
 
     def get_stack_change_points(self):
         # ref: ida.wll@0x100793d0
-        v = self.netnode.supval(tag='S', index=0x1000)
+        try:
+            v = self.netnode.supval(tag='S', index=0x1000)
+        except KeyError:
+            return
```
Comment (williballenthin) [inline idb/analysis.py]: i knew we'd have to transition towards this, thanks for helping out!

File: idb/analysis.py [hunk 2]
```diff
@@ -918,14 +920,14 @@ def get_stack_change_points(self):
 
 
 def _get_xrefs(db, tag, src=None, dst=None, types=None):
-    if not src and not dst:
+    if src is None and dst is None:
```
Comment (williballenthin) [inline idb/analysis.py]: good

File: idb/idapython.py [hunk 3]
```diff
@@ -552,32 +553,42 @@ def MaxEA(self):
         return segs[-1].endEA
 
     def GetFlags(self, ea):
-        return self.idb.id1.get_flags(ea)
+        try:
+            return self.idb.id1.get_flags(ea)
+        except KeyError:
+            return None
```
Comment (williballenthin) [inline idb/idapython.py]: thank you

File: idb/idapython.py [hunk 4]
```diff
@@ -631,189 +642,57 @@ def GetManyBytes(self, ea, size, use_dbg=False):
...
-        self.seg_dis = {}
-
+    def _load_dis(self, arch, mode):
         import capstone
+        if self.bit_dis is None:
+            self.bit_dis = {}
+        if self.bit_dis.get((arch, mode)) is None:
+            r = capstone.Cs(arch, mode)
+            self.bit_dis[(arch, mode)] = r
+        return self.bit_dis[(arch, mode)]
```
Comment (williballenthin) [inline idb/idapython.py]: this is a good enhancement, thanks

File: idb/idapython.py [hunk 5]
```diff
@@ -631,189 +642,57 @@ def GetManyBytes(self, ea, size, use_dbg=False):
...
 
     def _disassemble(self, ea):
+        import capstone
+
         size = self.ItemSize(ea)
-        buf = self.GetManyBytes(ea, size)
-        self._load_dis()
+        inst_buf = self.GetManyBytes(ea, size)
+        segment = self._get_segment(ea)
+        bitness = 16 << segment.bitness# 16, 32, 64
```
Comment (williballenthin) [inline idb/idapython.py]: much better than i was doing, thanks

File: idb/idapython.py [hunk 6]
```diff
@@ -631,189 +642,57 @@ def GetManyBytes(self, ea, size, use_dbg=False):
...
+        elif procname == "mips":
+            if bitness == 32:
+                dis = self._load_dis(capstone.CS_ARCH_MIPS, capstone.CS_MODE_MIPS32 | CS_MODE_BIG_ENDIAN)
+            elif bitness == 64:
+                dis = self._load_dis(capstone.CS_ARCH_MIPS, capstone.CS_MODE_MIPS64 | CS_MODE_BIG_ENDIAN)
+        elif procname == "mipsel":
+            if bitness == 32:
+                dis = self._load_dis(capstone.CS_ARCH_MIPS, capstone.CS_MODE_MIPS32 | CS_MODE_LITTLEN_ENDIAN)
+            elif bitness == 64:
+                dis = self._load_dis(capstone.CS_ARCH_MIPS, capstone.CS_MODE_MIPS64 | CS_MODE_LITTLEN_ENDIAN)
```
Comment (williballenthin) [inline idb/idapython.py]: excellent

**Comment by williballenthin** [comment]
Would you update the PR with the changes from @XVilka here: https://github.com/williballenthin/python-idb/pull/36/files

This should fix the travis build error so we can verify the unit tests.

**Comment by liumuqing** [comment]
Hi~
could you help to build idb for arm/mips binaries, It may not good for me to provide them.. you know.. due to the license/name info in the idb file.
I can send you some different arch ELF files and also help to write test cases in different arch if needed.

**Comment by williballenthin** [comment]
files received! added here: https://github.com/williballenthin/python-idb/tree/master/tests/data

Would you add a handful of tests to demonstrate that the disassembly on these architectures is working as expected? Then we can merge these changes.

Thank you!

**Comment by liumuqing** [comment]
of course~

File: idb/analysis.py (hunk 1 above)
Comment (liumuqing) [reply idb/analysis.py]: Since IDA's GetFlags return 0 in this case, maybe I should also return 0 here?

Comment (williballenthin) [reply idb/analysis.py]: yeah, we should try to be like IDA's API as close as possible. i was not doing a good job at this.

**Comment by williballenthin** [comment]
the new tests look really good. just what i had in mind.

**Comment by XVilka** [comment]
@williballenthin so, is it good for merging?

**Comment by williballenthin** [comment]
@XVilka done!

Major thanks to @liumuqing for making this happen!

**Pull request merged**